src/rspq/rspq.c

```diff
@@ -0,0 +1,1235 @@
+    return b;
+}
+
+void rspq_block_free(rspq_block_t *block)
+{
+    // Start from the commands in the first chunk of the block
+    int size = RSPQ_BLOCK_MIN_SIZE;
+    void *start = block;
+    uint32_t *ptr = block->cmds + size;
+    while (1) {
+        // Rollback until we find a non-zero command
+        while (*--ptr == 0x00) {}
+        uint32_t cmd = *ptr;
+
+        // If the last command is a JUMP
+        if (cmd>>24 == RSPQ_CMD_JUMP) {
+            // Free the memory of the current chunk.
+            free_uncached(start);
+            // Get the pointer to the next chunk
+            start = UncachedAddr(0x80000000 | (cmd & 0xFFFFFF));
+            if (size < RSPQ_BLOCK_MAX_SIZE) size *= 2;
+            ptr = (uint32_t*)start + size;
+            continue;
+        }
+        // If the last command is a RET
+        if (cmd>>24 == RSPQ_CMD_RET) {
+            // This is the last chunk, free it and exit
+            free_uncached(start);
+            return;
+        }
+        // The last command is neither a JUMP nor a RET:
+        // this is an invalid chunk of a block, better assert.
+        assertf(0, "invalid terminator command in block: %08lx\n", cmd);
+    }
+}
+
+void rspq_block_run(rspq_block_t *block)
+{
+    // TODO: add support for block execution in highpri mode. This would be
+    // possible by allocating another range of nesting levels (eg: 8-16) to use
+    // in highpri mode (to avoid stepping on the call stack of lowpri). This
+    // would basically mean that a block can either work in highpri or in lowpri
+    // mode, but it might be an acceptable limitation.
+    assertf(rspq_ctx != &highpri, "block run is not supported in highpri mode");
+
+    // Write the CALL op. The second argument is the nesting level
+    // which is used as stack slot in the RSP to save the current
+    // pointer position.
+    rspq_int_write(RSPQ_CMD_CALL, PhysicalAddr(block->cmds), block->nesting_level << 2);
+
+    // If this is CALL within the creation of a block, update
+    // the nesting level. A block's nesting level must be bigger
+    // than the nesting level of all blocks called from it.
+    if (rspq_block && rspq_block->nesting_level <= block->nesting_level) {
+        rspq_block->nesting_level = block->nesting_level + 1;
+        assertf(rspq_block->nesting_level < RSPQ_MAX_BLOCK_NESTING_LEVEL,
+            "reached maximum number of nested block runs");
+    }
+}
+
+void rspq_noop()
+{
+    rspq_int_write(RSPQ_CMD_NOOP);
+}
+
+rspq_syncpoint_t rspq_syncpoint(void)
+{   
+    assertf(!rspq_block, "cannot create syncpoint in a block");
+    assertf(rspq_ctx != &highpri, "cannot create syncpoint in highpri mode");
+    rspq_int_write(RSPQ_CMD_TEST_WRITE_STATUS, 
+        SP_WSTATUS_SET_INTR | SP_WSTATUS_SET_SIG_SYNCPOINT,
+        SP_STATUS_SIG_SYNCPOINT);
+    return ++rspq_syncpoints_genid;
+}
+
+bool rspq_check_syncpoint(rspq_syncpoint_t sync_id) 
+{
+    return sync_id <= rspq_syncpoints_done;
```

Comment:
I'd say it breaks briefly, and as soon as both `rspq_syncpoints_genid` and `rspq_syncpoints_done` have overflown it should go back to working correctly. `rspq_syncpoints_genid` will overflow first, which means during the period in which `rspq_syncpoints_done` hasn't, syncpoints that are created after the overflow will falsely be reported to have been reached.

Assuming we constantly generate one syncpoint every millisecond (at 60Hz that's 16 per frame), the overflow would happen after roughly 50 days of continuous uptime, based on my quick calculation. If that's is not acceptable we could make the syncpoint be 64 bits I guess?

Reply:
Consumers can also create syncpoints that might get triggerred many times a frame, so this might become a concern.
Still if we can verify it briefly breaks and then continue working, I think that is acceptable. Should be easy to test by starting at a high id initially.

Reply:
True, there isn't really an upper bound on how many syncpoints can be created. OTOH, we _do_ warn about abusing them in the docs :)

Jokes aside, even if they only break briefly that could have unforeseen and dramatic consequences for an application. So we should do our best to prevent that situation. @rasky It should be possible to bump the syncpoints IDs to 64 bit, right?

Reply:
With the 32-bit counter and 100 syncpoints per frame on a 60 fps application, we get an overflow after 7 days. This said, if we want to fix it, we can either switch to 64bit (I don't think there's any problem with that), or use a difference rather than a comparison to handle the overflow condition, just like `TICKS_BEFORE`. Using the latter would mean that the only limit is not to have more than 2 billion syncpoints which are enqueued and not yet executed, which I would say it's unlikely :)

Reply:
I like the difference method, which I hadn't even thought of. It guarantees syncpoints always working correctly, because it's not even physically possible to have that many commands in the queue. It's not even possible to have that many syncpoints in all of RDRAM with the expansion pack! So I'd say we do that?

Reply:
I mean it's also pretty much guaranteed that a 64 bit counter will never overflow, but it does involve doubling the storage size of syncpoints.

Reply:
Yeah, I think the difference is fine, it's easy enough to implement :)

Reply:
I've changed `rspq_check_syncpoint` to use the difference.